### ISSUE TITLE
Support sole tenancy

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#56993]
tl;dr: We need to modify  [PodAffinityTerm.Namespaces](https://github.com/kubernetes/api/blob/5cb170de1f060ebf2dde62cc4b8241fdb3c01203/core/v1/types.go#L2584) to allow a way to say "all namespaces." We also need an optional admission controller to add a "tenant" label to all non-system pods.

Full details:

[Taints and tolerations](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/) enable "dedicated nodes" i.e. statically space-partition your cluster among users/workloads, although this isn't yet secure (see https://github.com/kubernetes/kubernetes/issues/48041#issuecomment-323266418 and https://github.com/kubernetes/community/pull/911). 

A more efficient way of ensuring that untrusting/conflicting tenants don't share a node, is "sole tenancy." In this model a pod can only use a node if it is empty (ignoring system pods) or is only running pods from the same tenant (ignoring system pods). This dynamically space-partitions the cluster among tenants and doesn't require the management overhead of dedicated nodes. (Of course in both approaches, some nodes can be shared, by tenants that have pods that don't need exclusive access, if that's desired.) Sole tenancy is what this issue is about.

Unfortunately there's no way to do this right now.

To understand the rest of this issue, you need to understand [PodAffinityTerm](https://github.com/kubernetes/api/blob/5cb170de1f060ebf2dde62cc4b8241fdb3c01203/core/v1/types.go#L2584) which is the way a pod that is requesting pod affinity or anti-affinity specifies which pods it wants to affinitize/anti-affinitize to/from. The important part is that it takes a list of namespaces, and a label selector that is applied against the pods in that/those namespace(s). If the list of namespaces is empty, it means "my namespace." There is no way to specify "all namespaces except mine" (you'd have to list them individually, and keep them up-to-date). Label selector is described [here](https://github.com/kubernetes/apimachinery/blob/a04e753f5223cf882db01ac64212682ea28767b4/pkg/apis/meta/v1/types.go#L899); empty label selector matches all objects. Lastly, note that you can specify more than one pod affinity and/or pod anti-affinity expression for a pod, and the resulting sets of nodes are intersected. (There is no way to do OR.) 

First, note that hard pod affinity (e.g. the tenant labels all of its pods with some label tenant=X, and uses hard pod affinity to say "only run me on nodes that are already running pods with tenant=X") doesn't work because (1) the scheduler may choose a node that is already running pods without tenant=X, and (2) even if the scheduler picks an initially empty node, once it fills up that node with tenant=X pods, the rest of the tenant's pods will go pending. (If you're wondering how the first pod from the tenant gets scheduled, see the special case code [here](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L1305).) An alternative way of using hard pod affinity has the same problem: don't bother with the labels, and say "only run me on nodes that have pods that match the 'return all objects' label selector in my namespace" (i.e. on nodes that have pods from my namespace). Note that there is no way to say "OR empty nodes" for either of these approaches, but such a feature wouldn't help anyway because it would only solve problem (2) not problem (1).

So, we need to use hard pod anti-affinity. What we want to say is "do NOT run me on nodes that are already running a pod from any namespace such that the pod has label tenant != X" (note that "do NOT run me on nodes that are already running pods from another namespace" sounds appealing but doesn't work because nodes invariably will have some pods from the kube-system namespace running). Unfortunately you can't express this today because PodAffinityTerm has no way to say "all namespaces" -- you either have to give an explicit list of namespaces you care about or say "my namespace." So we need to change `PodAffinityTerm.Namespaces` to let you say "all namespaces." (I believe it just requires changing [this function](https://github.com/kubernetes/kubernetes/blob/078a3b653b9c1bd74adb4265f3bc8a3dbc386153/plugin/pkg/scheduler/algorithm/priorities/util/topologies.go#L28).) Then we use an admission controller to label all pods with a "tenant" label, and a pod from tenant X that wants sole tenancy can put in its PodSpec a PodAntiAffinity that looks like (pseudo-config here!)
```
PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution = {
    [LabelSelector: tenant NOT IN {"X"}
    Namespaces: all namespaces
    TopologyKey: node]
}
```

Per-node system pods (e.g. fluentd) would not have a tenant label, otherwise no sole-tenant pods would ever schedule. Cluster-level system pods (e.g. heapster) could or could not have a tenant label, depending on whether the cluster admin wants sole tenant pods to be able to co-schedule with cluster-level system pods. Note that the reason this works (i.e. a pod with with such an anti-affinity rule can schedule onto a node with pods that have no "tenant" label) is that `tenant NOT IN {"X"}` will not match pods that lack the tenant label (due to [this code](https://github.com/kubernetes/apimachinery/blob/18a564baac720819100827c16fdebcadb05b2d0d/pkg/labels/selector.go#L187)).

To make this secure, the admission controller that adds a "tenant" label to all non-system pods should also deny creation of pods that already have the label if the value is not set properly (or change it to the proper one), and deny updating a pod to change or remove the "tenant" label.

BTW I'm 90% sure we originally let you say "all namespaces" but removed that feature for some reason. But I don't see how to do sole tenancy without it. It would be good to re-discover why we removed it, though (probably performance issues?). There is hopefully some way we can make this not have terrible performance (e.g. do the label comparison once per pod instead of once per pod per namespace in the system).

The idea of having the system automatically add a "namespace" label to every pod has been discussed in the past. I believe it was in the discussion about how to select pods for NetworkPolicy.

/kind feature
/sig scheduling
/sig auth

**Comment by davidopp** [comment]
cc/ @kubernetes/sig-scheduling-feature-requests

**Comment by davidopp** [comment]
cc/ @kevin-wangzefeng

**Comment by wojtek-t** [comment]
I think we added (but looking into code apparently it never happened) an ability to specify:
`NamespaceSelector` instead of explicit list of namespaces in PodAffinityTerm.

But since it seems it never happened, I think that's the way how we should proceed with it.

**Comment by davidopp** [comment]
Oh right, I totally forgot about it. You're thinking of https://github.com/kubernetes/kubernetes/pull/54536

**Comment by kevin-wangzefeng** [comment]
> BTW I'm 90% sure we originally let you say "all namespaces" but removed that feature for some reason. But I don't see how to do sole tenancy without it. It would be good to re-discover why we removed it, though (probably performance issues?). There is hopefully some way we can make this not have terrible performance (e.g. do the label comparison once per pod instead of once per pod per namespace in the system).

We originally introduced the semantic "all namespaces" by using nil list of `Namespaces []string`, while empty list means "all namespaces". In the initial version we didn't add json tag "omitempty" for `Namespaces` since we need to distinguish nil and empty. IIRC, it was removed since it doesn't work (probably after protobuf encode and decode).
